Sources/Paywalls/Events/PaywallEventStore.swift

```diff
@@ -0,0 +1,162 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  PaywallEventStore.swift
+//
+//  Created by Nacho Soto on 9/5/23.
+
+import Foundation
+
+protocol PaywallEventStoreType {
+
+    /// Stores `event` into the store.
+    @available(iOS 15.0, tvOS 15.0, macOS 12.0, watchOS 8.0, *)
+    func store(_ storedEvent: PaywallStoredEvent) async
+
+    /// - Returns: the first `count` events from the store.
+    @available(iOS 15.0, tvOS 15.0, macOS 12.0, watchOS 8.0, *)
+    func fetch(_ count: Int) async -> [PaywallStoredEvent]
+
+    /// Removes the first `count` events from the store.
+    @available(iOS 15.0, tvOS 15.0, macOS 12.0, watchOS 8.0, *)
+    func clear(_ count: Int) async
+
+}
+
+@available(iOS 15.0, tvOS 15.0, macOS 12.0, watchOS 8.0, *)
+internal actor PaywallEventStore: PaywallEventStoreType {
+
+    private let handler: FileHandlerType
+
+    init(handler: FileHandlerType) {
+        self.handler = handler
+    }
+
+    func store(_ storedEvent: PaywallStoredEvent) async {
+        do {
+            Logger.verbose(PaywallEventStoreStrings.storing_event(storedEvent.event))
+
+            await self.handler.append(line: try PaywallEventSerializer.encode(storedEvent))
+        } catch {
+            Logger.error(PaywallEventStoreStrings.error_storing_event(error))
+        }
+    }
+
+    func fetch(_ count: Int) async -> [PaywallStoredEvent] {
+        assert(count > 0, "Invalid count: \(count)")
+
+        do {
+            return try await self.handler.readLines()
+                .prefix(count)
+                .compactMap { try? PaywallEventSerializer.decode($0) }
+                .extractValues()
+        } catch {
+            Logger.error(PaywallEventStoreStrings.error_fetching_events(error))
+            return []
+        }
+    }
+
+    func clear(_ count: Int) async {
+        assert(count > 0, "Invalid count: \(count)")
+
+        do {
+            try await self.handler.removeFirstLines(count)
+        } catch {
+            Logger.error(PaywallEventStoreStrings.error_removing_first_lines(count: count, error))
+
+            // If removing these `count` events fails, try removing the entire file.
+            // This ensures that we don't try to send the same events again.
```

Comment:
maybe this should be a docstring for the method instead? feels like a pretty important detail

Reply:
👍🏻